File: libzebra/src/main/cpp/utils/ByteUtil.cpp
//
// Created by Administrator on 2022/2/12.
//

#include "ByteUtil.h"
#include <cstring>

int64_t ByteUtil::sysIdToInt64(char *data, int len) {
    int64_t tid = 0;
    for (int i = 0; i < len; i++) {
        int64_t ret = 0;
        char x = data[i] & 0xFF;
        if (x >= '0' && x <= '9') {
            ret = x - '0';
        } else if (x >= 'A' && x <= 'Z') {
            ret = x - 'A' + 10;
        } else if (x >= 'a' && x <= 'z') {
            ret = x - 'a' + 10;
        } else {
            continue;
        }
        int64_t pow = 1;
        int32_t y = len - i;
        for (int i = 1; i < y; i++) {
            pow = pow * 36;
        }
        tid += ret * pow;
    }
    return tid;
}

void ByteUtil::int64ToSysId(char *data, int64_t tid) {
    int64_t num = tid;
    int len = 0;
    char temp[16];
    while (num > 36) {
        int8_t ni = (num % 36) & 0xFF;
        if (ni < 10) {
            temp[len++] = ni + '0';
        } else {
            temp[len++] = ni - 10 + 'A';
        }
        num = num / 36;
    }
    if (num < 10) {
        temp[len] = num + '0';
    }
    else {
        temp[len] = num - 10 + 'A';
    }
    for (int i = 0; i <= len; i++) {
        data[i] = temp[len - i];
    }
}

int16_t ByteUtil::getInt16(const char *data) {
    return (int16_t) (data[0] & 0xFF) << 8 | (data[1] & 0xFF);
}

void ByteUtil::int16ToData(char *data, int16_t i1, int16_t i2, int16_t i3, int16_t i4) {
    data[0] = i1 >> 8 & 0xFF;
    data[1] = i1 & 0xFF;
    if (i2 == 0) return;
    data[2] = i2 >> 8 & 0xFF;
    data[3] = i2 & 0xFF;
    if (i3 == 0) return;
    data[4] = i3 >> 8 & 0xFF;
    data[5] = i3 & 0xFF;
    if (i4 == 0) return;
    data[6] = i4 >> 8 & 0xFF;
    data[7] = i4 & 0xFF;
}

int32_t ByteUtil::getInt32(const char *data) {
    return (int32_t) (data[0] & 0xFF) << 24 | (data[1] & 0xFF) << 16 | (data[2] & 0xFF) << 8 |
           (data[3] & 0xFF);
}

void ByteUtil::int32ToData(char *data, int32_t i1, int32_t i2) {
    data[0] = i1 >> 24 & 0xFF;
    data[1] = i1 >> 16 & 0xFF;
    data[2] = i1 >> 8 & 0xFF;
    data[3] = i1 & 0xFF;
    if (i2 == 0) return;
    data[4] = i2 >> 24 & 0xFF;
    data[5] = i2 >> 16 & 0xFF;
    data[6] = i2 >> 8 & 0xFF;
    data[7] = i2 & 0xFF;
}

int16_t ByteUtil::byte2int16(const char *data, int32_t offset, bool littleEndian) {
    int16_t value = 0;
    for (int count = 0; count < 2; ++count) {
        int shift = (littleEndian ? count : (1 - count)) << 3;
        value |= ((int16_t) 0xff << shift) & ((int16_t) data[offset + count] << shift);
    }
    return value;
}

void ByteUtil::int16ToBytes(int16_t value, char *data, int offset, bool littleEndian) {
    for (int count = 0; count < 2; ++count) {
        int shift = (littleEndian ? count : (1 - count)) << 3;
        data[count + offset] = (char) (value >> shift & 0xFF);
    }
}

int32_t ByteUtil::byte2int32(const char *data, int32_t offset, bool littleEndian) {
    int32_t value = 0;
    for (int count = 0; count < 4; ++count) {
        int shift = (littleEndian ? count : (3 - count)) << 3;
        value |= ((int32_t) 0xff << shift) & ((int32_t) data[offset + count] << shift);
    }
    return value;
}

void ByteUtil::int32ToBytes(int32_t value, char *data, int offset, bool littleEndian) {
    for (int count = 0; count < 4; ++count) {
        int shift = (littleEndian ? count : (3 - count)) << 3;
        data[count + offset] = (char) (value >> shift & 0xFF);
    }
}

int64_t ByteUtil::byte2int64(const char *data, int32_t offset, bool littleEndian) {
    int64_t value = 0;
    for (int count = 0; count < 8; ++count) {
        int shift = (littleEndian ? count : (7 - count)) << 3;
        value |= ((long) 0xff << shift) & ((long) data[offset + count] << shift);
    }
    return value;
}

void ByteUtil::int64ToBytes(int64_t value, char *data, int offset, bool littleEndian) {
    for (int count = 0; count < 8; ++count) {
        int shift = (littleEndian ? count : (7 - count)) << 3;
        data[count + offset] = (char) (value >> shift & 0xFF);
    }
}

File: libutils/src/main/java/com/flyzebra/utils/ByteUtil.java
package com.flyzebra.utils;

/**
 * Author: FlyZebra
 * Created by flyzebra on 18-4-9-下午3:11.
 */
public class ByteUtil {
    public static String bytes2HexString(byte[] bytes) {
        if (bytes == null || bytes.length == 1) {
            return null;
        }
        StringBuilder sb = new StringBuilder("");
        for (byte aByte : bytes) {
            String hv = Integer.toHexString(aByte & 0xFF);
            if (hv.length() < 2) {
                sb.append(0);
            }
            sb.append(hv);
            sb.append(":");
        }
        return sb.toString().toUpperCase();
    }

    public static String bytes2HexString(byte[] bytes, int length) {
        if (bytes == null || length < 1) {
            return "";
        }
        StringBuilder sb = new StringBuilder("");
        length = Math.min(length, bytes.length);
        for (int i = 0; i < length; i++) {
            String hv = Integer.toHexString(bytes[i] & 0xFF);
            if (hv.length() < 2) {
                sb.append(0);
            }
            sb.append(hv);
            sb.append(":");
        }
        return sb.toString().toUpperCase();
    }

    public static String bytes2HexString(byte[] bytes, int start, int length) {
        if (bytes == null || length < 1 || bytes.length < (start + length)) {
            return "";
        }
        StringBuilder sb = new StringBuilder("");
        length = Math.min(start + length, bytes.length);
        for (int i = start; i < length; i++) {
            String hv = Integer.toHexString(bytes[i] & 0xFF);
            if (hv.length() < 2) {
                sb.append(0);
            }
            sb.append(hv);
            sb.append(":");
        }
        return sb.toString().toUpperCase();
    }

    public static byte[] hexString2Bytes(String hex) {
        int len = (hex.length() / 2);
        byte[] result = new byte[len];
        char[] achar = hex.toCharArray();
        for (int i = 0; i < len; i++) {
            int pos = i * 2;
            result[i] = (byte) (toByte(achar[pos]) << 4 | toByte(achar[pos + 1]));
        }
        return result;
    }

    private static byte toByte(char c) {
        return (byte) "0123456789ABCDEF".indexOf(c);
    }

    public static short bytes2Short(byte[] bytes, int offset, boolean littleEndian) {
        short value = 0;
        for (int count = 0; count < 2; ++count) {
            int shift = (littleEndian ? count : (1 - count)) << 3;
            value |= ((short) 0xff << shift) & ((short) bytes[offset + count] << shift);
        }
        return value;
    }

    public static void shortToBytes(short value, byte[] bytes, int offset, boolean littleEndian) {
        for (int count = 0; count < 2; ++count) {
            int shift = (littleEndian ? count : (1 - count)) << 3;
            bytes[count + offset] = (byte) (value >> shift & 0xFF);
        }
    }

    public static int bytes2Int(byte[] bytes, int offset, boolean littleEndian) {
        int value = 0;
        for (int count = 0; count < 4; ++count) {
            int shift = (littleEndian ? count : (3 - count)) << 3;
            value |= ((int) 0xff << shift) & ((int) bytes[offset + count] << shift);
        }
        return value;
    }

    public static void intToBytes(int value, byte[] bytes, int offset, boolean littleEndian) {
        for (int count = 0; count < 4; ++count) {
            int shift = (littleEndian ? count : (3 - count)) << 3;
            bytes[count + offset] = (byte) (value >> shift & 0xFF);
        }
    }

    public static long bytes2Long(byte[] bytes, int offset, boolean littleEndian) {
        long value = 0;
        for (int count = 0; count < 8; ++count) {
            int shift = (littleEndian ? count : (7 - count)) << 3;
            value |= ((long) 0xff << shift) & ((long) bytes[offset + count] << shift);
        }
        return value;
    }

    public static void longToBytes(long value, byte[] bytes, int offset, boolean littleEndian) {
        for (int count = 0; count < 8; ++count) {
            int shift = (littleEndian ? count : (7 - count)) << 3;
            bytes[count + offset] = (byte) (value >> shift & 0xFF);
        }
    }

    public static String longToSysId(long tid) {
        long num = tid;
        int len = 0;
        byte[] temp = new byte[16];
        while (num > 36) {
            byte ni = (byte) ((num % 36) & 0xFF);
            if (ni < 10) {
                temp[len++] = (byte) (ni + '0');
            } else {
                temp[len++] = (byte) (ni - 10 + 'A');
            }
            num = num / 36;
        }
        if (num < 10) {
            temp[len] = (byte) (num + '0');
        } else {
            temp[len] = (byte) (num - 10 + 'A');
        }
        byte[] data = new byte[len + 1];
        for (int i = 0; i <= len; i++) {
            data[i] = temp[len - i];
        }
        return new String(data);
    }

    public static long sysIdToInt64(String stid) {
        long tid = 0;
        byte[] data = stid.getBytes();
        for (int i = 0; i < data.length; i++) {
            long ret = 0;
            byte x = (byte) (data[i] & 0xFF);
            if (x >= '0' && x <= '9') {
                ret = x - '0';
            } else if (x >= 'A' && x <= 'Z') {
                ret = x - 'A' + 10;
            } else if (x >= 'a' && x <= 'z') {
                ret = x - 'a' + 10;
            } else {
                continue;
            }
            long pow = 1;
            int y = data.length - i;
            for (int j = 1; j < y; j++) {
                pow = pow * 36;
            }
            tid += ret * pow;
        }
        return tid;
    }
}
